guava/src/com/google/common/collect/ImmutableSet.java

```diff
@@ -722,20 +739,92 @@ static int maxRunBeforeFallback(int tableSize) {
       this.hashCode = toCopy.hashCode;
+
+      if (toCopy.overflowTables != null) {
+        int size = toCopy.overflowTables.size();
+        this.overflowTables = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+          Object[] table = toCopy.overflowTables.get(i);
+          this.overflowTables.add(Arrays.copyOf(table, table.length));
+        }
+      }
     }
 
+    @SuppressWarnings("unchecked")
     void ensureTableCapacity(int minCapacity) {
       if (minCapacity > expandTableThreshold && hashTable.length < MAX_TABLE_SIZE) {
-        int newTableSize = hashTable.length * 2;
-        hashTable = rebuildHashTable(newTableSize, dedupedElements, distinct);
+        /**
+         * we have about {@code (a_0 sum_{k=1 to n+1} of 2^k)} elements total in our table, assuming
+         * we fill each of the overflow tables fully, which is {@code a_0 * (2^{n+2} - 2) <=
+         * hashtable.size * 2} by geometric series.
+         *
+         * <p>We always allocate enough to fill all elements in our table to needing to rehash
+         * frequently.
+         */
+        int newTableSize = chooseTableSize(hashTable.length * 2);
+        if (overflowTables == null) {
+          overflowTables = new ArrayList<>();
+        }
+        overflowTables.add(dedupedElements);
+        distinct = 0;
+        dedupedElements = (E[]) new Object[expandTableThreshold];
+
+        numElementsInHashTable = 0;
+        hashTable = new Object[newTableSize];
+
+        hashCode = 0;
         maxRunBeforeFallback = maxRunBeforeFallback(newTableSize);
         expandTableThreshold = (int) (DESIRED_LOAD_FACTOR * newTableSize);
       }
     }
 
+    @SuppressWarnings("unchecked")
+    void flattenOverflowTables() {
+      if (overflowTables == null) {
+        return;
+      }
+
+      // ensure capacity first to avoid resizing and inserting into a high load table as much as
+      // possible
+      ensureTableCapacity(hashTable.length * 2);
+
+      ArrayList<Object[]> overflows = overflowTables;
+      overflowTables = null;
+      // now we rehash.
+      for (int i = 0; i < overflows.size(); i++) {
+        Object[] table = overflows.get(i);
+        for (int j = 0; j < table.length && table[j] != null; j++) {
+          add((E) table[j]);
+        }
+      }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    SetBuilderImpl<E> combine(SetBuilderImpl<E> other) {
+      if (other instanceof RegularSetBuilderImpl) {
+        RegularSetBuilderImpl<E> otherR = (RegularSetBuilderImpl<E>) other;
+        SetBuilderImpl<E> result = this;
+        for (int i = 0; i < other.distinct; i++) {
+          result = result.add(other.dedupedElements[i]);
+        }
+        if (otherR.overflowTables != null) {
+          for (int i = 0; i < otherR.overflowTables.size(); i++) {
+            Object[] tables = otherR.overflowTables.get(i);
+            for (int j = 0; j < tables.length && tables[j] != null; j++) {
```

Comment:
The overflow tables are arrays of deduped elements (per each of the hash sets), which are not necessarily filled full (and in fact most often not filled completely full).

Reply:
I think they should be, I guess?  Here's what I thought your implementation was going for, every time a reset occurs:

- dedupedElements is set to a new, empty array of size k (which is twice the old length of dedupedElements)
- hashTable is set to a new array of size chooseTableSize(k)
- each time an element is added, we add it to the *current* hash table, and if it's not a duplicate of something already in there, it gets added to dedupedElements
- we reset when dedupedElements is full, meaning the hash table is saturated to its intended capacity

If that's not what you were going for, can you explain what you meant instead?

Reply:
In your outlined case, because `chooseTableSize(k)` returns powers of 2s, we end up wasting some capacity of the hashTable.
Since rehashes are the most expensive, I chose to maximize the use of the hashTable instead.

That is on the reset:
* hashTable is set to a new array of size chooseTableSize(currentHashTableCap * 2)
* dedupedElements is set to new array of size `expandTableThreshold`.
* we reset when `expandTableThreshold` is met, that is table is saturated.

corollary is that `dedupedElements` is full, expect that the initial construction of the hashTable and `dedupedElements` don't follow these conditions. I could modify the constructor to follow these conditions as well, guaranteeing max fill.

However, the flattened result of `dedupedElements` is not max filled. It is sized according to the max possible size resulting from the flattening of all overflows, but there could be some duplicates between the overflows.
While not the case in this implementation, my next patch with some fixes makes it clear that when we copy builders, we can not guaranty that `dedupedElements` are filled.
Furthermore, the constructor of `RegularSetBuilderImpl` doesn't guarantee the sizes that result in always filled conditions either.